Enable minion's IPC channel to aggregate results from spawned jobber processes

devkits opened the pull request:
### What does this PR do?

Enable minion's IPC channel to aggregate results from spawned jobber processes. Use a long-running request channel in the minion parent process to communicate job results back to the master via broker-based or broker-less transport.

This is a necessary optimization at scale for transports that prefer a sustained long-running connection because connection create/dispose operations are expensive. The working assumption is that this change benefits all supported transports.

Testing Done:
 * this tests provide coverage for this use case:
      .../salt/tests/pytests/integration/minion.*

### Previous Behavior
Minion would spawn a process that would run a job and create/dispose a new transport connection just to communicate one job result. This does not scale well for transports that prefer persistent connections because connection operations are expensive.

### New Behavior
Minion would spawn a process that would run a job and communicate result via an IPC channel to the parent minion process. Parent minion process would use a long-running connection to communicate result back to master. This approach scales better because it reduces connection churn when thousands of minions are responding to the master. 


Fixes: #61274

devkits committed:
Enable minion's IPC channel to aggregate results from spawned jobber processes. Use a long-running request channel in the minion parent process to communicate job results back to the master via broker-based or broker-less transport.

This is a necessary optimization for transports that prefer a sustained long-running connection because connection create/dispose operations are expensive. The working assumption is that this change benefits all supported transports.

Testing Done:
 * this tests provide coverage for this use case:
      .../salt/tests/pytests/integration/minion.*

devkits committed:
Fix minion unit tests, specifically .../tests/pytests/test_minion.py

devkits committed:
Merge branch 'master' into minion_job_completion_optimization

devkits commented:
re-run full all

jfindlay commented on salt/minion.py:
```diff
@@ -1364,11 +1364,41 @@ def connect_master(self, failed=False):
         """
         Return a future which will complete when you are connected to a master
         """
+        # Consider refactoring so that eval_master does not have a subtle side-effect on the contents of the opts array
```
Someday, `opts` (`__opts__`, `__salt__`, *et al*) will be readonly, rite?

dwoz replied:
I'm not sure how quickly that could happen. In any case, we want to tamper with `__opts__` as little as possible.

jfindlay commented:
You might consider also updating https://github.com/saltstack/salt/blob/master/doc/topics/development/architecture.rst#minion-job-flow.

devkits commented:
> You might consider also updating https://github.com/saltstack/salt/blob/master/doc/topics/development/architecture.rst#minion-job-flow.

Thanks. I will update https://github.com/saltstack/salt/blob/master/doc/topics/development/architecture.rst#minion-job-flow shortly.

devkits commented:
re-run full all

devkits committed:
Reuse minion's existing local IPC bus to fire job completion and other events intended for the master's ReqServer. Tag these specific events with a specific tag "__master_req_channel_payload"
Update the architecture flow accordingly: https://github.com/saltstack/salt/blob/master/doc/topics/development/architecture.rst#minion-job-flow

Testing Done:
  - updated unit tests,
  - tested manually with zeromq

devkits committed:
Merge branch 'master' into minion_job_completion_optimization

devkits committed:
Merge branch 'master' into minion_job_completion_optimization

dwoz commented:
@devkits [This test](https://jenkins.saltproject.io/job/pr-centos-7-x86_64-py3-tcp-pytest/job/PR-61468/8/testReport/tests.pytests.integration.minion/test_return_retries/test_publish_retry/) needs to be fixed. Otherwise, the PR looks good.

devkits committed:
Merge branch 'master' into minion_job_completion_optimization

devkits commented:
> 

Thanks. I can reproduce the test failure locally. Looks like the fix may need  be in the TCP transport itself. I will try to update the PR with the fix.

dwoz committed:
Move retries to channel

dwoz committed:
Clean up cruft

dwoz committed:
Pre commit fixes

garethgreenaway commented:
@devkits Can you please add a changelog for this PR?  Thanks!

dwoz committed:
Add changelog
